### PR TITLE
DOC: Fix flake8 problems in io.rst

### LINKED ISSUE #23791: DOC: Fix format of io.rst
Our documentation should follow the PEP-8 standard, but for historical reasons there are many formats that don't. We should fix them now.

This issue is to fix the format of `doc/source/io.rst`. To find the errors, first is required to install a newer version of `flake8-rst`:
```
conda install -c conda-forge flake8-rst=0.5.0
```

Then, all the errors can be obtained with:
```
$ flake8-rst doc/source/io.rst 
doc/source/io.rst:14:4: E402 module level import not at top of file
doc/source/io.rst:19:4: E402 module level import not at top of file
doc/source/io.rst:22:4: E402 module level import not at top of file
doc/source/io.rst:430:88: E501 line too long (84 > 79 characters)
doc/source/io.rst:481:4: E402 module level import not at top of file
doc/source/io.rst:666:17: E126 continuation line over-indented for hanging indent
doc/source/io.rst:680:10: E222 multiple spaces after operator
doc/source/io.rst:721:83: E501 line too long (95 > 79 characters)
doc/source/io.rst:815:10: E222 multiple spaces after operator
doc/source/io.rst:898:4: E402 module level import not at top of file
doc/source/io.rst:995:9: E111 indentation is not a multiple of four
doc/source/io.rst:1019:83: E501 line too long (87 > 79 characters)
doc/source/io.rst:1020:83: E501 line too long (89 > 79 characters)
doc/source/io.rst:1021:83: E501 line too long (95 > 79 characters)
doc/source/io.rst:1036:10: E222 multiple spaces after operator
doc/source/io.rst:1135:10: E222 multiple spaces after operator
doc/source/io.rst:1147:12: E222 multiple spaces after operator
doc/source/io.rst:1169:9: E225 missing whitespace around operator
doc/source/io.rst:1440:55: E231 missing whitespace after ','
doc/source/io.rst:1455:4: E402 module level import not at top of file
doc/source/io.rst:1483:4: E402 module level import not at top of file
doc/source/io.rst:1970:13: E128 continuation line under-indented for visual indent
doc/source/io.rst:1971:13: E128 continuation line under-indented for visual indent
doc/source/io.rst:2023:12: E999 SyntaxError: invalid syntax
doc/source/io.rst:2062:4: E402 module level import not at top of file
doc/source/io.rst:2075:4: E402 module level import not at top of file
doc/source/io.rst:2084:14: E127 continuation line over-indented for visual indent
doc/source/io.rst:2086:19: E126 continuation line over-indented for hanging indent
doc/source/io.rst:2088:14: E127 continuation line over-indented for visual indent
doc/source/io.rst:2089:26: E128 continuation line under-indented for visual indent
doc/source/io.rst:2090:26: E128 continuation line under-indented for visual indent
doc/source/io.rst:2091:13: E127 continuation line over-indented for visual indent
doc/source/io.rst:2094:19: E126 continuation line over-indented for hanging indent
doc/source/io.rst:2149:8: E124 closing bracket does not match visual indentation
doc/source/io.rst:2189:6: E402 module level import not at top of file
doc/source/io.rst:2325:4: E402 module level import not at top of file
doc/source/io.rst:2823:26: E231 missing whitespace after ':'
doc/source/io.rst:2823:44: E231 missing whitespace after ':'
doc/source/io.rst:2824:66: E231 missing whitespace after ','
doc/source/io.rst:2845:83: E501 line too long (80 > 79 characters)
doc/source/io.rst:2853:4: E402 module level import not at top of file
doc/source/io.rst:3175:4: E402 module level import not at top of file
doc/source/io.rst:3252:4: E402 module level import not at top of file
doc/source/io.rst:3309:41: E201 whitespace after '{'
doc/source/io.rst:3309:50: E202 whitespace before '}'
doc/source/io.rst:3424:4: E402 module level import not at top of file
doc/source/io.rst:3438:30: E231 missing whitespace after ','
doc/source/io.rst:3453:4: E402 module level import not at top of file
doc/source/io.rst:3462:28: E127 continuation line over-indented for visual indent
doc/source/io.rst:3467:28: E127 continuation line over-indented for visual indent
doc/source/io.rst:3481:13: E128 continuation line under-indented for visual indent
doc/source/io.rst:3482:13: E128 continuation line under-indented for visual indent
doc/source/io.rst:3484:37: E225 missing whitespace around operator
doc/source/io.rst:3485:12: E128 continuation line under-indented for visual indent
doc/source/io.rst:3486:12: E126 continuation line over-indented for hanging indent
doc/source/io.rst:3588:30: E241 multiple spaces after ','
doc/source/io.rst:3663:38: E231 missing whitespace after ':'
doc/source/io.rst:3670:52: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3670:54: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3823:83: E501 line too long (81 > 79 characters)
doc/source/io.rst:3865:4: E402 module level import not at top of file
doc/source/io.rst:3866:30: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3866:32: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3866:60: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3866:62: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3866:64: E201 whitespace after '['
doc/source/io.rst:3866:83: E501 line too long (140 > 79 characters)
doc/source/io.rst:3866:140: E202 whitespace before ']'
doc/source/io.rst:3950:45: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3950:47: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:3979:7: E111 indentation is not a multiple of four
doc/source/io.rst:4006:9: E111 indentation is not a multiple of four
doc/source/io.rst:4006:23: E226 missing whitespace around arithmetic operator
doc/source/io.rst:4011:9: E111 indentation is not a multiple of four
doc/source/io.rst:4108:38: E127 continuation line over-indented for visual indent
doc/source/io.rst:4113:66: E202 whitespace before '}'
doc/source/io.rst:4114:30: E127 continuation line over-indented for visual indent
doc/source/io.rst:4123:30: E127 continuation line over-indented for visual indent
doc/source/io.rst:4123:38: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:4123:40: E251 unexpected spaces around keyword / parameter equals
doc/source/io.rst:4162:46: E202 whitespace before ')'
doc/source/io.rst:4335:49: E202 whitespace before '}'
doc/source/io.rst:4349:4: E402 module level import not at top of file
doc/source/io.rst:4427:29: E127 continuation line over-indented for visual indent
doc/source/io.rst:4438:4: E402 module level import not at top of file
doc/source/io.rst:4526:4: E402 module level import not at top of file
doc/source/io.rst:4592:4: E402 module level import not at top of file
doc/source/io.rst:4676:4: E402 module level import not at top of file
doc/source/io.rst:4725:84: E501 line too long (85 > 79 characters)
doc/source/io.rst:4746:4: E402 module level import not at top of file
doc/source/io.rst:4805:4: E402 module level import not at top of file
doc/source/io.rst:4836:4: E402 module level import not at top of file
doc/source/io.rst:4838:36: E231 missing whitespace after ','
doc/source/io.rst:4838:39: E231 missing whitespace after ','
doc/source/io.rst:4839:4: E128 continuation line under-indented for visual indent
doc/source/io.rst:4839:31: E231 missing whitespace after ','
doc/source/io.rst:4839:34: E231 missing whitespace after ','
doc/source/io.rst:4840:4: E122 continuation line missing indentation or outdented
doc/source/io.rst:4840:31: E231 missing whitespace after ','
doc/source/io.rst:4840:34: E231 missing whitespace after ','
doc/source/io.rst:4842:8: E221 multiple spaces before operator
doc/source/io.rst:4873:5: E402 module level import not at top of file
doc/source/io.rst:5002:84: E501 line too long (81 > 79 characters)
doc/source/io.rst:5057:4: E402 module level import not at top of file
doc/source/io.rst:5067:8: E128 continuation line under-indented for visual indent
doc/source/io.rst:5074:65: E712 comparison to True should be 'if cond is True:' or 'if cond:'
doc/source/io.rst:5080:5: E402 module level import not at top of file
doc/source/io.rst:5242:4: E402 module level import not at top of file
```

After fixing this issue, the previous command should not report any error.

### PR DESCRIPTION
**Pull request by saurav2608**
- [x] closes #23791
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Pushing this code to check if this clears CI. Made a few changes towards issue #23791

### CODE REVIEW

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: It's somehow unrelated to your changes, but if you don't mind removing this line, and replacing `ExcelWriter` by `pd.ExcelWriter` wherever is used, that would be great. I think reassigning this just make things difficult for the users to understand.

And if `import pandas.util.testing as tm` is used just in one code block or section, would be nice to move it there, as it may not be obvious for users to know what `tm` is without the import. If it's used in many places leave it here, and we'll see what we can do later on.

**Review by datapythonista — APPROVED** [review]
lgtm, but there are couple of unrelated things, that if you can fix here, that would be great.

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: those lines are not equivalent

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: I think this should be left where it was. Any reason to move it here?

or even better, I'd use `pd..io.date_converters...`, no extra import is needed, and the code is more explicit

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: are those needed for the validation? I'd say flake8-rst should fail with 3 blank lines

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: I think it'll be more readable as:
```
data = b'...'
data = data.decode(...
```

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: I don't think this new blank line is correct

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: same, blank line not needed

**Review by datapythonista — CHANGES_REQUESTED** [review]
Couple of comments

**Comment by datapythonista** [comment]
@saurav2608 in general try to not open the pull requests until you're done. You can push to your remote branch, but if you open a PR, someone will review it, and it's a waste of time if your changes are not finished (unless you need a review to make sure what you're doing is correct...)

**Comment by saurav2608** [comment]
> @saurav2608 in general try to not open the pull requests until you're done.

@datapythonista -  I will keep this in mind henceforth. I will edit the .rst file over the weekend.

**Comment by datapythonista** [comment]
@saurav2608 do you have time to address the comments?

**Comment by saurav2608** [comment]
I got swamped by some other work. I can take this up on Wednesday.

File: doc/source/io.rst
Comment (saurav2608) [reply doc/source/io.rst]: I am unable to resolve an error here. On my local build the the second line is not appended to the variable data. I am quoting the code here. Need you help. 
`
In [68]: data = '# empty\n# second empty line\n# third empty' \

In [69]: 'line\nX,Y,Z\n1,2,3\nA,B,C\n1,2.,4.\n5.,NaN,10.0'
Out[69]: 'line\nX,Y,Z\n1,2,3\nA,B,C\n1,2.,4.\n5.,NaN,10.0'

In [70]: print(data)
\# empty
\# second empty line
\# third empty

In [71]: pd.read_csv(StringIO(data), comment='#', skiprows=4, header=1)
---------------------------------------------------------------------------
EmptyDataError                            Traceback (most recent call last)
<ipython-input-71-2ad3ec79b56f> in <module>
----> 1 pd.read_csv(StringIO(data), comment='#', skiprows=4, header=1)
`

Comment (saurav2608) [reply doc/source/io.rst]: Note that if I copy paste these lines to a ipython terminal. It works well. I tried combination of spaces etc.

Comment (saurav2608) [reply doc/source/io.rst]: @datapythonista - Can you please confirm the above? Otherwise everything is ready for a commit and push.

Comment (datapythonista) [reply doc/source/io.rst]: Sorry, I missed that.

I didn't quite understand what's the problem, but checking it again, I think much better ways to write that are:
```python
data = '\n'.join(['# empty',
                  '# second empty line',
                  '# third empty',
                  'line',
                  'X,Y,Z',
                  '1,2,3',
                  'A,B,C',
                  '1,2.,4.',
                  '5.,NaN,10.0'])
```
or
```python
data = '''# empty
# second empty line
# third empty
line
X,Y,Z
1,2,3
A,B,C
1,2.,4.
5.,NaN,10.0'''
```

which makes possible to read the content

**Comment by saurav2608** [comment]
HI @datapythonista - travis fails with the below error. 

> [create env]
Solving environment: ...working... failed
>ResolvePackageNotFound: 
>  - flake8-rst=0.4.2

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: sorry I didn't see it before, but looks like mostly everywhere on this file, what is being used is:
```
data = ('#empty\n'
        '#second empty line'
...
```
I think it would be nice to keep the same for consistency. Or was this the format that was failing in this case?

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: do you mind breaking this in lines in the way previously suggested?

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: same here

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: same

**Review by datapythonista — COMMENTED** [review]
I rerun travis. But reviewing this again, I saw that we're using different ways of creating multiline strings. If you don't mind, would be great to standardize them

**Comment by jreback** [comment]
can you rebase

**Comment by saurav2608** [comment]
> can you rebase

@jreback : Unfortunately, I am not able to. I messed up at some point and now rebase and squash generates conflict errors that I am not resolve it with my knowledge of git. 

I am open to ideas on how to squash the commit. In the worst case, I can delete the branch and start again.

**Comment by datapythonista** [comment]
just do a `git fetch upstream && git merge upstream/master` edit the file with the conflict, keep the lines that arr ok, and leave the file how it has to be, do a `git add` of the file, commit and push

**Comment by saurav2608** [comment]
I am not sure if this worked.

**Comment by saurav2608** [comment]
> I am not sure if this worked.

It did not. Somehow an old version of the file was pushed.

**Comment by datapythonista** [comment]
Can you take a look at the first section of: https://datapythonista.github.io/blog/useful-git-commands.html

**Comment by saurav2608** [comment]
@datapythonista - I followed your post. But if I try to push I get below error. Should I pull one before pushing? 
`  ! [rejected]            io-doc -> io-doc (non-fast-forward)
error: failed to push some refs to 'https://github.com/saurav2608/pandas.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.`

**Comment by saurav2608** [comment]
> @datapythonista - I followed your post. But if I try to push I get below error. Should I pull one before pushing?
> ` ! [rejected] io-doc -> io-doc (non-fast-forward) error: failed to push some refs to 'https://github.com/saurav2608/pandas.git' hint: Updates were rejected because the tip of your current branch is behind hint: its remote counterpart. Integrate the remote changes (e.g. hint: 'git pull ...') before pushing again. hint: See the 'Note about fast-forwards' in 'git push --help' for details.`

This is done I think. I forced a push on my forked repo.

**Review by datapythonista — APPROVED** [review]
lgtm, thanks @saurav2608

**Comment by datapythonista** [comment]
@saurav2608 can you do in this PR branch: `git fetch upstream && git merge upstream/master && git push` please

I'm not sure why the CI is failing, I think it shouldn't, but hopefully updating the branch will fix it.

**Comment by datapythonista** [comment]
@saurav2608 looks like the PR contains unrelated changes again. Not sure what is causing this, I don't think the command I wrote causes it. But in any case, can you repeat what you did before to fix the history and leave only your changes in the PR please.

**Comment by saurav2608** [comment]
I redid the PR. this look okay now. Please review once and let me know.

File: doc/source/io.rst
Comment (datapythonista) [reply doc/source/io.rst]: any reason to not split this in several lines as in the rest?

File: doc/source/io.rst
Comment (datapythonista) [inline doc/source/io.rst]: I guess this is giving an error, but that will be fixed in flake8-rst, as the error is wrong. So just one blank line please.

**Review by datapythonista — COMMENTED** [review]
thanks @saurav2608, just couple of small things if you don't mind

**Review by datapythonista — APPROVED** [review]
lgtm

thanks a lot for fixing all these @saurav2608

**Comment by saurav2608** [comment]
@datapythonista  - no problem. It took long for my to fix this. Thank for your guidance. There are still some issues to resolve. For example suppressing `E402 module level import not at top of file` type errors where the import is deliberately kept at that location. I can work on those once these changes are merged.

**Comment by datapythonista** [comment]
No worries @saurav2608, we all do what we can, and it also takes time for us to review.

I don't think we want to fix the `E402`, but make `flake8-rst` not report them as errors. I think it's better for users to have the imports in the block where they are used.

And don't worry much if not everything is fixed. There are so many things to fix at this point, that any fixes are very useful. Once most of the stuff is fixed, then we'll check more in detail the outstanding problems.

If you're looking for other things to contribute, it's probably better to wait until #23847 is merged to work on the rest of the `.rst` files, as it's a bit difficult to follow right now what is done, what's in progress, and what needs to be done. After that PR is merged, I'll create issues for the pages that need fixes, and it'll be easier.

One related issue that can be address and would be very useful to have implemented soon would be #23952.

**Comment by jreback** [comment]
@saurav2608 can you rebase and remove io.rst from the setup.cfg

**Comment by datapythonista** [comment]
@jreback if you want to merge this, I'll open a PR with the fixed files, and I'll add this one too.

**Comment by jreback** [comment]
thanks @saurav2608 

@datapythonista go for it

**Pull request merged**
